File: leetcode/longestSubstring.cpp
/**********************************************
  > File Name		: longestSubstring.cpp
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Sat 27 Feb 2021 04:43:14 PM CST
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

class Solution {
public:
    int longestSubstring(string s, int k) {
        int ret = 0;
        int n = s.length();
        for (int t = 1; t <= 26; t++) {
            int l = 0, r = 0;
            vector<int> cnt(26, 0);
            int tot = 0;
            int less = 0;
            while (r < n) {
                cnt[s[r] - 'a']++;
                if (cnt[s[r] - 'a'] == 1) {
                    tot++;
                    less++;
                }
                if (cnt[s[r] - 'a'] == k) {
                    less--;
                }

                while (tot > t) {
                    cnt[s[l] - 'a']--;
                    if (cnt[s[l] - 'a'] == k - 1) {
                        less++;
                    }
                    if (cnt[s[l] - 'a'] == 0) {
                        tot--;
                        less--;
                    }
                    l++;
                }
                if (less == 0) {
                    ret = max(ret, r - l + 1);
                }
                r++;
            }
        }
        return ret;
    }
};

//作者：LeetCode-Solution
//链接：https://leetcode-cn.com/problems/longest-substring-with-at-least-k-repeating-characters/solution/zhi-shao-you-kge-zhong-fu-zi-fu-de-zui-c-o6ww/
//来源：力扣（LeetCode）
//著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

File: leetcode/longest_substring.rs
/**********************************************
  > File Name		: longest_substring.rs
  > Author		    : lunar
  > Email			: lunar_ubuntu@qq.com
  > Created Time	: Sat 27 Feb 2021 03:07:49 PM CST
  > Location        : Shanghai
  > Copyright@ https://github.com/xiaoqixian
 **********************************************/

/*
 * Leetcode 395: find the longest substring, and every character exists in the substring has to
 * repeat at least k times.
 *
 * In Rust, string processing is always a tricky problem.
 * We can first count the number of times all characters appear in the string.
 * Then we just need to find a substring with all characters repeat more than k times.
 */

use std::collections::HashMap;

struct Solution {}

use std::cmp;

impl Solution {
    pub fn longest_substring(s: String, k: i32) -> i32 {
        let mut res = 0;
        let len = s.len();
        let chars: Vec<char> = s.chars().collect();
        for t in 1..27 {
            let mut l: usize = 0;
            let mut r: usize = 0;
            let mut count: Vec<i32> = vec![0; 26];
            let mut tot = 0;
            let mut less = 0;
            while r < len {
                count[(chars[r] as u32 - 'a' as u32) as usize] += 1;
                if count[(chars[r] as u32 - 'a' as u32) as usize] == 1 {
                    tot += 1;
                    less += 1;
                }
                if count[(chars[r] as u32 - 'a' as u32) as usize] == k {
                    less -= 1;
                }
                while tot > t {
                    count[(chars[l] as u32 - 'a' as u32) as usize] -= 1;
                    if count[(chars[l] as u32 - 'a' as u32) as usize] == k - 1 {
                        less += 1;
                    }
                    if count[(chars[l] as u32 - 'a' as u32) as usize] == 0 {
                        tot -= 1;
                        less -= 1;
                    }
                    l += 1;
                }
                if less == 0 {
                    res = cmp::max(res, r as i32 - l as i32 + 1);
                }
                r += 1;
            }
        }
        res
    }
}


fn main() {
    Solution::longest_substring(String::from("aaaccsdsd"), 2);
}
